### DISCUSSION TITLE
Feedback on a possible Alire Unicode policy

### DISCUSSION THREAD

**Discussion by mosteo** [alire-project/alire#1334] [Polls]
We are again discussing what (or if) Alire should do something about Unicode. My last attempt at documenting the situation is at https://github.com/alire-project/alire/pull/1332 (see [unicode.md](https://github.com/alire-project/alire/blob/d2a75519e792b66d38d0eefc24267c148b7ae888/doc/unicode.md) and [AEP-0004.md](https://github.com/alire-project/alire/blob/d2a75519e792b66d38d0eefc24267c148b7ae888/doc/AEPs/aep-0004.md) for full details). In a way I see this as the same pain that arose in the Python2 -> Python3 transition.

**I want to get more feedback from experts and the community feeling.** There's a poll below.

The gist of the conumdrum is:

- By default GNAT presumes Latin-1 sources.
- With `-gnatW8` GNAT presumes UTF-8 sources and changes `Text_IO` and `Wide_Wide_Text_IO` default behavior *globally*.
- Regular specs and generic bodies are recompiled from different contexts and may end compiled in both modes.
    - String literals may end in memory with different encodings, sometimes causing compilation errors.
 - `Ada.Text_IO` will behave differently depending on `-gnatW8` being active.
    - Without `-gnatW8`, `Ada.Text_IO` prints bytes as-is. 
    - With `-gnatW8`, `Ada.Text_IO` applies a Latin1 --> UTF8 conversion which garbles UTF-8 strings.
    - An unfortunate consequence is that  `Ada.Text_IO` works "as expected" for UTF-8 sources *without* `-gnatW8`.
- `Ada.Wide_Wide_Text_IO` doesn't work by default without `-gnatW8` for non-Latin1 characters (prints brackets encoding).

There's no good default that will work for all cases for both unaware and aware users, which is the main point of contention.

Reading past debates at `comp.lang.ada` (mainly [this one](https://groups.google.com/g/comp.lang.ada/c/l0aflBePm_E/)) it seems this is beyond repair from the ARG side, and there are proponents of both extremes, even of avoiding it completely by staying in ASCII in sources.

In the poll below you'll find the following options:

- Leave things as they are (default to no `-gnatW8`). 
   - This is the Alire 1.x behavior, each crate for itself.
   - Crates trying to do "the right thing" (using `-gnatW8`) are more likely to find trouble from non-gnatW8 crates.
   - Literals and `Text_IO` don't really behave as they should, even if this counterintuitively works.
- Make `-gnatW8` the default (this is currently enabled in the `master` branch)
   - It may cause errors when trying to use some crates already indexed.
     - A few don't compile due to literals, which is easy to detect and fix.
     - The impact on non-gnatW8 crates that do compile and use `Ada.Text_IO` is unclear until runtime.
   - People may be surprised with `Ada.Text_IO` garbling UTF-8 strings.
   - It may be a headache of unknown proportions but eventually everyone is on the same page (hopefully?).
- Some other option, please contribute your thinking in the comments.

**Comment by mgrojo** [comment]
I know everybody uses UTF-8 locales these days, but, at least on GNU/Linux, wouldn't it make sense to switch the behaviour depending on the value of the `LANG` environment variable? For example, when `LANG` contains UTF-8, then `-gnatW8`. Otherwise, no `-gnatW8`. I don't know if there are equivalent locale configuration in other operating systems to do similar conditionals.

**Reply by mosteo** [reply]
Unfortunately, sources are tied to the `-gnatW8` status, so they cannot be compiler indistinctly with/without.

**Reply by simonjwright** [reply]
macOS (in the UK):
```
$ echo $LANG
en_GB.UTF-8
```

**Reply by mgrojo** [reply]
> Unfortunately, sources are tied to the `-gnatW8` status, so they cannot be compiler indistinctly with/without.

I don't follow you. What I meant is: default to UTF-8 when Alire crate is initialized if LANG contains UTF-8. Default to do nothing otherwise. What you are proposing is determined at crate initialization or at crate building?

**Reply by mosteo** [reply]
Sorry, I misunderstood you. What you propose makes sense in isolation, but the problem for `Alire` is that mixing crates with and without `-gnatW8` may lead to both compilation errors and unexpected `Ada.Text_IO` behavior. So for crates intended for publishing that's where things get dicey.

**Reply by mgrojo** [reply]
Yes, you're right, I only gave a thought to an isolated development, but when you publish your sources, the value of your particular `LANG` variable does no longer matter at all.

**Comment by jquorning** [comment]
Disclaimer: I do not know much about encoding and codepages and such.

But the discussion rang a bell about databases;
When setting up MS SQL Server you can select something about encoding/codepages and such.
(I just use the default and hope for the best)
Can we learen some lessons from the database community perhaps?

**Reply by mosteo** [reply]
If you could narrow down a bit the information to look at, I'd definitely will give it a read.

More generally, I think the general computing trend has been for years to steadily move towards adopting Unicode, and UTF-8 as the default Unicode encoding and environment. Trouble is always the breakage along the way.

Mind you, for Alire I don't think the situation is as dire as it sounds. It's only to agree on a couple of things, and for most code it won't be even a point to consider. The problem IMO is the huge inertia with using `Ada.Text_IO`.

**Reply by jquorning** [reply]
`Ada.Text_IO` is for `ASCII` or perpaps Latin-1 or whathavewenot..

A better way IMO would be something like
`package Project.Text_IO is ...`  
`package UTF8.Text_IO is ...`
`package Unicode.Text_IO is ...`

;and then perhaps
`package Text_IO renames Ada.Text_IO`
`package Text_IO renames ...`
`package IO renames ...`

YMMV

**Reply by Micronian2** [reply]
Hi, I don’t normally use Text_IO, so I haven’t had to deal with the issues. However, given the bullet item

“ With -gnatW8, Ada.Text_IO applies a Latin1 --> UTF8 conversion which garbles UTF-8 strings” 

Does this mean Ada.Text_IO should no longer be used if -gnatW8 is applied, and instead Ada.Wide_Text_IO should be used instead?

**Reply by mosteo** [reply]
That's it, @Micronian2.

**Reply by OCTAGRAM** [reply]
> Hi, I don’t normally use Text_IO, so I haven’t had to deal with the issues. However, given the bullet item
> 
> “ With -gnatW8, Ada.Text_IO applies a Latin1 --> UTF8 conversion which garbles UTF-8 strings”
> 
> Does this mean Ada.Text_IO should no longer be used if -gnatW8 is applied, and instead Ada.Wide_Text_IO should be used instead?

Text_IO is trash and Wide_Wide_Text_IO is not much better. I think about borrowing [TTextWriter](https://docwiki.embarcadero.com/Libraries/Alexandria/en/System.Classes.TTextWriter) from Delphi, able to operate on any stream, unlike Ada IO. Modula-2 has some useful stuff too, to some extent. I like [ReadToken](https://www.modula2.org/reference/isomodules/isomodule.php?m=STEXTIO.DEF). Delphi's TTextReader/TTextWriter with Modula-2 addons would be nice.

Also terminal handling stuff is useful: moving cursor over screen, and setting colors like Turbo Pascal was able to do since 1980s. And waiting for "single keypress".

I almost find no use of Ada's paragraph stuff occuping so much space in Ada.(Wide_Wide_)Text_IO.

**Comment by joakim-strandberg** [comment]
I voted for UTF8 support for all Ada sources in Alire. But it's not without pause. Personally, I stick to ASCII only in my personal projects. It is my impression that one of the original design goals of the Ada language is for the code to look as much as possible like normal English prose. Just look at all the reserved keywords, they are all English words. Thus, introducing the possibility of using UTF8 in the source code with the purpose of the code looking like English prose doesn't seem like a natural course of action. Being able to use identifiers in the source code expressed in Spanish, Japanese or Farsi in one and the same code base won't help maintainability. Well, it is possible to put Farsi in the source code even when the code is ASCII only, for example "function Payam return String;". From the compilers perspective, in general, the identifiers only need to be unique no matter what language is used. Let's say it is possible to use a static code analysis tool to check that an Ada code base only uses identifiers from English regardless of the Ada source code is in ASCII or UTF8. Why else would one want to use UTF8 in the code base? Perhaps one would like to put Emojis in the comments of the source code. Perhaps that is the main motivation for allowing UTF8 in one's source code? No, thinking about it, I would say the main motivation for UTF8 in the source code is to be able to embed UTF8 strings. When sticking to ASCII only, all the UTF8 strings need to be put in separate files from the Ada source code. It is more convenient to be able to embed UTF8 strings directly in the source code. So there is a use-case where it is more convenient, but not necessary.

Another reason I avoid UTF8 in the source code are the existence of different code-points in UTF8 which look the same (https://util.unicode.org/UnicodeJsps/confusables.jsp?a=test&r=None) . I don't want the possibility in my personal projects to define a function Hello and then having the compiler tell me it is not possible to call it because there is no function Hello, and the difference would be perhaps for the "e" in the definition of the function and the "e" used in the function call to be different code points but look the same to the naked eye. It would not take me long to figure out what's wrong, but it is a source of error that I am happy to be without.

Yet another reason against UTF8 Ada source code is security. Allowing UTF8 in the source code enables dangerous UTF8 character sequences (https://krebsonsecurity.com/2021/11/trojan-source-bug-threatens-the-security-of-all-code/). When for example writing Go code there is a static source analysis tool for detecting these in Go source code: https://github.com/breml/bidichk
Even if one has a static code analysis tool to detect dangerous UTF8 character sequences, how can I be sure the tool detects all possible dangerous character sequences? Well, since I can't convince myself of it I would need to make a white list of allowed words (character sequences) to be used in the Ada source code. Sticking to ASCII I avoid this problem completely and seem to align with the Keep It Simple principle.

The main motivation for allowing UTF8 is to make the transition for newcomers to the Ada language easier. If people have the culture and tradition of putting UTF8 in the source code of their Go, Rust, Java, C# applications etc. it may be off-putting if the Ada source code would be ASCII only.

**Reply by Micronian2** [reply]
Interesting. I never considered a lot of UTF8 in the source code itself.  I assumed the greater need is to process UTF8 data, especially data that is produced from a third party.

**Reply by mosteo** [reply]
Yeah, for me the issue is only to allow string literals in sources. The identifiers should preferably remain ASCII. Unfortunately, even the ARM uses the Pi example. I guess in some very limited cases like math it may improve readability, but the drawbacks seem overpowering.

**Reply by jquorning** [reply]
I guess string literals are for toy projects like *Hello, World!* anyway. With internationalization string literals are best put into a database (See *WinCC*, the *Siemens* SCADA IDE).

**Reply by stcarrez** [reply]
> I guess string literals are for toy projects like _Hello, World!_ anyway. With internationalization string literals are best put into a database (See _WinCC_, the _Siemens_ SCADA IDE). Maybe we could learn some lessons from eastern part of Europe (read: Russia, artillery 6:1)

I have a counter example with [Ada Stemmer](https://github.com/stcarrez/ada-stemmer) library where Snowball generates stems in literal strings. The stemmer supports 27 languages including [Russian](https://github.com/stcarrez/ada-stemmer/blob/master/src/lang/stemmer-russian.adb), [Greek](https://github.com/stcarrez/ada-stemmer/blob/master/src/lang/stemmer-greek.adb) and [Arabic](https://github.com/stcarrez/ada-stemmer/blob/master/src/lang/stemmer-arabic.adb). Storing those strings in database is not practical. Having them as UTF-8 in the source gives a readable code and helps.

The stemmer algorithm works only with UTF-8 strings.

**Reply by mgrojo** [reply]
[Ada Stemmer](https://github.com/stcarrez/ada-stemmer) is then a good example of a library that would be broken when using `-gnatW8`. @mosteo, would it be impossible to use this library with Alire, or after configuration would be possible, provided that any other Alire library you include is ASCII only? I personally still haven't got a clear idea of how much pain this will bring.

**Reply by stcarrez** [reply]
No, I'm not sure it will be broken since it does not use `Ada.Text_IO` at all. I suspect the main issue lies in reading and writing UTF-8 from. If you compile the library, within the library strings are encoded as sequence of bytes in UTF-8 form. The question is when you have to write those strings externally or read some and you are using one of the `Ada.*Text_IO` packages.

To build [Ada Stemmer](https://github.com/stcarrez/ada-stemmer) I don't use the `-gnatW8`.

Now, with [Ada Wiki](https://github.com/stcarrez/ada-wiki) that's a different story and I had to use the `-gnatW8` to compile the samples despite the fact that the library and its samples do not have any UTF-8 static strings. However, the library is only based on `Wide_Wide_Character` for parsing Wiki syntax and so on. The issue comes when you read/write something outside/inside the Ada/GNAT world with the `Ada.Wide_Wide_Text_IO` package.

The root issue is that `Ada.Wide_Wide_Text_IO` is broken. At least, it should be configurable so that the encoding be specified when reading and writing text. But I guess that's another issue.

**Reply by mgrojo** [reply]
I thought those string literals outside the Latin-1 set would raise some of the compilation errors I saw in [aep-0004.md#impacted-crates](https://github.com/alire-project/alire/blob/d2a75519e792b66d38d0eefc24267c148b7ae888/doc/AEPs/aep-0004.md#impacted-crates), like
```
asf-converters-numbers.adb:131:53: error: literal out of range of type Standard.Character
excel_writer=17.0.1 excel_out_demo.adb:125:39: illegal wide character
```

**Reply by reznikmm** [reply]
String is NOT a sequence of bytes, it's a sequence of Latin-1 characters. See ARM 3.5.2 2/3:

> The predefined type Character is a character type whose values correspond to the 256 code points of Row 00 (also known as Latin-1) of the ISO/IEC 10646:2011 Basic Multilingual Plane (BMP).

For sure Ada Stemmer contains incorrect Ada code. For example
```ada
   Among_String : constant String := "вшись" &
```

Without `-gnatW8` the compiler read `ш` as 0xD1 0x88 Latin-1 characters. where 0xD1 is `Ñ`, but 0x88 is the control character *Character (Horizontal) Tabulation Set - HTS* and should be rejected during compilation according to ARM 2.1 4/3:

> The only characters allowed outside of comments are those in categories *other_format, format_effector, and graphic_character*.

**Reply by stcarrez** [reply]
> String is NOT a sequence of bytes, it's a sequence of Latin-1 characters. See ARM 3.5.2 2/3:
> 

Thank you Maxime for the clarification.

I've always considered String as a sequence of bytes due to the `UTF_String` type being mapped to String which I understand is a mistake.

**Reply by stcarrez** [reply]
The consequence of Ada String being Latin1 means that `Ada.Text_IO` has to make the UTF-8 conversion.
The following piece of code:

```Ada
with Ada.Strings.UTF_Encoding.Wide_Wide_Strings;
with Ada.Text_IO;
procedure Print is
  S : String := Ada.Strings.UTF_Encoding.Wide_Wide_Strings.Encode ("вшись");
begin
  Ada.Text_IO.Put_Line (S);
end Print;
```

will never print the expected value despite the fact that at some point `S` is a valid encoded UTF-8 sequence.

Does this mean that if we want to support UTF-8 in some program, we must not use `Ada.Text_IO` ?

**Reply by OCTAGRAM** [reply]
Your projects that you are putting into Alire, are then going to be linked with the main program that is not supposed to be uploaded to Alire, but this main program can contain a lot of native strings because the program only makes sense in a specific locale. For instance, EAEU customs formats handling only makes sense in one country only due to different local laws.

Also if the whole developer team are not native English, but program is internationalized, then developers write strings in the language they know best, and then dedicated translators provide English counterparts on demand. I worked in several such projects. What is the point in requiring all developers to have good written English knowledge if we don't sell much to English speaking countries?

**Comment by stcarrez** [comment]
The `-gnatW8` is broken and I think we should not use it. Here is why.

Consider the following simple Ada program that reads the current directory and print file names. This Ada program only uses the ASCII character set, no UTF-8 in it:

```Ada
with Ada.Directories;
with Ada.Text_IO;
procedure Readdir is
   Search_Filter : constant Ada.Directories.Filter_Type
     := (Ada.Directories.Ordinary_File => True,
         Ada.Directories.Directory     => True,
         Ada.Directories.Special_File  => False);
   Search   : Ada.Directories.Search_Type;
   Ent      : Ada.Directories.Directory_Entry_Type;
begin
   Ada.Directories.Start_Search (Search, Directory => ".",
                                 Pattern => "*", Filter => Search_Filter);
   while Ada.Directories.More_Entries (Search) loop
      Ada.Directories.Get_Next_Entry (Search, Ent);
      Ada.Text_Io.Put_Line (Ada.Directories.Simple_Name (Ent));
   end loop;
end Readdir;
```

Create a file that shows the issue:

```
touch Adélaïde
```

If you compile it with `gprbuild readdir.adb` (no additional option), the program itself is able to print UTF-8 filenames correctly. You will see: `Adélaïde`.

If you compile it with `gprbuild readdir.adb -gnatW8`, then an UTF-8 filename is printed with garbage.
You will see: `AdÃ©laÃ¯de`.

The same issue will arise if you get your UTF-8 string from a database or from an i18N database such as `gettext (3)`.

Please, stay away from that broken compiler option until it is really fixed.

**Reply by mgrojo** [reply]
That is one of the cases where Text_IO output is garbled. What is disappointing in this case is that you haven't read the UTF-8 string from a third library, but only using the Ada standard library. I agree with you that this is a bug, if Text_IO output applies Latin-1 to UTF-8 conversions with `-gnatW8`, then `Ada.Directories` should apply the reverse conversion when reading from the OS (UTF-8 to Latin-1). Then the drawback of `-gnatW8` would be performance degradation, assumed in order to be standard compliant. Another problem is what to do when the input cannot be represented as Latin-1. Would be acceptable that GNAT garbles the input or raises an exception?

Everything seems confusing, because the [documentation of `-gnatW8`](https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gnat_ugn/Character-Set-Control.html#Character-Set-Control) only mentions that it affects the encoding of Wide_Character; and not how it affects all these standard String interfaces.

Note that in Ada 2022 Wide and Wide_Wide versions of `Ada.Directories` have been added to cope with Unicode filenames. Using those and `Wide_Wide_Text_IO`, the program should work.

I've taken a look at http://www.ada-auth.org/cgi-bin/cvsweb.cgi/ai12s/ai12-0021-1.txt?rev=1.4 and https://groups.google.com/g/comp.lang.ada/c/s4ZUgrIBV5s/m/LesChmRjqHUJ and what can I only say that's there seems not to be an easy solution.

**Reply by mgrojo** [reply]
Bad news, `Ada.Wide_Wide_Directories` is still not implemented, at least in FSF GNAT 12.
```
$ gprbuild -cargs -gnat2022 wide_wide_readdir.adb
using project file /usr/share/gpr/_default.gpr
Compile
   [Ada]          wide_wide_readdir.adb
wide_wide_readdir.adb:1:06: error: "Ada.Wide_Wide_Directories" is not a predefined library unit
wide_wide_readdir.adb:1:06: error: "Ada.Wide_Wide_Directories" is not a predefined library unit
gprbuild: *** compilation phase failed
```

**Reply by simonjwright** [reply]
Neither Wide_Directories nor Wide_Wide_Directories are in GCC 13.0.1 (20230226)

**Reply by Fabien-Chouteau** [reply]
> The `-gnatW8` is broken and I think we should not use it.

That's my feeling as well. Any idea how to fix it or provide a better version?

**Reply by jquorning** [reply]
I guess a proper terminal driver is needed;

package IO renames Terminal.Text_IO

with

package Terminal renames Terminal_Driver_Text_IO

Or the like
YMMV

**Reply by OCTAGRAM** [reply]
Propert `gettext` binding shall isolate from its implementation details like NUL-termination and/or byte-encoding by converting from/to UTF-32. In Delphi 2009 (which was 14 years ago) bindings are done like that, except Delphi stuck in '90s Unicode UTF-16.

Some Linux APIs like libidn2 can accept UTF-32 directly.

**Reply by stcarrez** [reply]
> Propert `gettext` binding shall isolate from its implementation details like NUL-termination and/or byte-encoding by converting from/to UTF-32. In Delphi 2009 (which was 14 years ago) bindings are done like that, except Delphi stuck in '90s Unicode UTF-16.

The GtkAda library uses `gettext` and maps the result to a `UTF8_String` which is declared as subtype of String. I guess it works because all those strings will be processed sooner or later by Gtk which considers strings as sequence of bytes. But, if we try to print the `gettext` string with `Ada.Text_IO`, it will not work. The Wide version of String would solve this but at the expense of increased complexity in that library I suspect.

**Reply by OCTAGRAM** [reply]
> > Propert `gettext` binding shall isolate from its implementation details like NUL-termination and/or byte-encoding by converting from/to UTF-32. In Delphi 2009 (which was 14 years ago) bindings are done like that, except Delphi stuck in '90s Unicode UTF-16.
> 
> The GtkAda library uses `gettext` and maps the result to a `UTF8_String` which is declared as subtype of String. I guess it works because all those strings will be processed sooner or later by Gtk which considers strings as sequence of bytes. But, if we try to print the `gettext` string with `Ada.Text_IO`, it will not work. The Wide version of String would solve this but at the expense of increased complexity in that library I suspect.

Then GtkAda binding is wrong. They should have used Wide_Wide_String.

I have faced GPS crashes on cyrillic letters because of this. GPS handled Ada fine, but crashed on non-Ada source.

Ada application crashing. Pathetic.

Stop this madness already.

**Reply by mosteo** [reply]
> GPS handled Ada fine, but crashed on non-Ada source.

I have had not crashes but bizarre behavior with Spanish-heavy UTF-8 sources. On-the-fly casing is broken (it capitalizes after a non-ASCII letter), which causes style warnings and with warnings-as-errors it's a pain. Tabulation is also broken, wild guessing because there are some mixups of bytes/character counts.

**Reply by simonjwright** [reply]
> The `-gnatW8` is broken and I think we should not use it. Here is why.

No such bad behaviour if you use `GNAT.IO`

**Comment by reznikmm** [comment]
ARM [2.1 16](http://www.ada-auth.org/standards/rm12_w_tc1/html/RM-2-1.html) says:
> An Ada implementation shall accept Ada source code in UTF-8 encoding

Without `-gnatW8` GNAT is NOT compliant with ARM.

**Reply by simonjwright** [reply]
Looking at [ACATS test C250002](https://github.com/simonjwright/ACATS/blob/master/tests/c2/c250002.au), judiciously edited for reasons documented in [GCC PR81114](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81114) to remove the use of library unit names using extended characters (e.g. `C250002_Á`), we have source code in UTF-8 which is accepted (by FSF GCC 12.2.0) without the use of `-gnatW8`.

The fact that the source has a BOM (which `gnatchop` handles! without `-r`, anyway) probably has a lot to do with it.

In any case, whether you need `-gnatW8` or not, GNAT _does_ accept source in UTF-8.

**Reply by reznikmm** [reply]
Here is my cont-example. Save it in UTF-8 and compile:
```ada
procedure Main is
   C : Character := '©';
begin
   null;
end Main;
```

    $ gcc -c main.adb 
    main.adb:2:21: error: strings are delimited by double quote character

So, my conclusion is that the GNAT doesn't accept source in UTF-8 without `-gnatW8` (and BOM) and GNAT without `-gnatW8` is NOT compliant with ARM. ARM [2.1 16](http://www.ada-auth.org/standards/rm12_w_tc1/html/RM-2-1.html) says:

> An Ada implementation shall accept Ada source code in UTF-8 encoding, with or **without a BOM**...

PS `C250002` contains BOM, that's why GNAT passes it.

**Reply by simonjwright** [reply]
Just looked at [AI05-0286](http://www.ada-auth.org/cgi-bin/cvsweb.cgi/ai05s/ai05-0286-1.txt?rev=1.7&raw=N) and you are of course right.

From that document,
>An Ada implementation shall accept Ada source code in UTF-8 encoding, with or without a BOM ...

>AARM Reason: This is simply requiring that an Ada implementation be able to directly process the ACATS, which is provided in the described format. ...

I feel quite strongly that this is a case of the tail wagging the dog!

**Comment by stcarrez** [comment]
As @reznikmm pointed out, the Ada String is a sequence of Latin-1 which to my understanding means that:

* We must use the `Wide_Text_IO` or `Wide_Wide_Text_IO` for a correct UTF-8 support,
* We should convert with `Decode` any String that is obtained from some C-level API to the Wide form prior using it (if it uses UTF-8 encoding, `gettext (3)`, `dcgettext (3)` come to my mind),
* We can use the `Ada.Directories` on UTF-8 filenames but we have to `Decode` the string to the Wide form
* We could use Ada String (or best UTF8_String) to hold a sequence of bytes as long as we manage the input/output ourselves,
* We shall use Wide_String or Wide_Wide_String to declare a constant string initialized with charsets outside of Latin-1

Any other advises?

**Reply by OCTAGRAM** [reply]
We shall do something about exceptions too. They store 200 bytes, in UTF-8 this is likely to be nearly 100 characters and broken UTF-8 sequences appear constantly

**Reply by mosteo** [reply]
> Any other advises?

I'd say that's the gist of it. Though it's unclear to me if the `Ada.Directories` non-wide are supposed to return Latin1 or just the raw OS-provided strings. I hope the ARM says something about it.

**Reply by mgrojo** [reply]
> I'd say that's the gist of it. Though it's unclear to me if the `Ada.Directories` non-wide are supposed to return Latin1 or just the raw OS-provided strings. I hope the ARM says something about it.

If `Ada.Text_IO` is able to output Latin-1 in a UTF-8 environment, I don't see a reason why `Ada.Directories` is not returning Latin-1 from that UTF-8 environment. OK, there will be a penalty from doing the translation back and forth, when compared to the default mode, which does nothing. But once you've assumed that `Ada.Text_IO` has to convert from Latin-1 due to the environment working in UTF-8, the opposite direction should do the opposite conversion, IMHO.

In fact, there's something that I don't understand in the `-gnatW8` mode, I don't know if I'm missing something, or the implementation is not complete or inconsistent. The switch is probably doing more than it should do, but at the same time, it has only gone half way. There should be a switch to indicate how is the source code encoded, without any translation on run time, thus if you `Text_IO.Put` a Latin-1 literal, it would be garbled, because it has been read from the source code correctly and stored as Latin-1 at build time. But if you get some UTF-8 somewhere and print it, it would display right, just by chance and out of any Ada legality, like in the default mode. If you want the literal case to work, then you have to change your terminal settings to Latin-1, the same as if you encode your source in Latin-1 and print using `Text_IO.Put` (and you cannot blame GNAT for not following the ARM in that case; it's just that the OS has moved away from 8-bit character encodings). This switch wouldn't have any penalty on runtime, it's just for informing the compiler how have you encoded your sources, it's just a compilation time matter.

Then, if you want the compiler to automagically convert Latin-1 to UTF-8 at run time, so you can use `Text_IO.Put` without having to tweak your terminal, then a separate switch should be used. This switch should work independently of how your sources are encoded. That is, separate switches for translations in compilation time and in run time should exist, and they would be perfectly compatible. It should make sense to encode your sources in Latin-1 and, at the same time, to enable this switch so Latin-1 characters work in a UTF-8 environment, assuming the overhead.

Am I misunderstanding something?

**Reply by jquorning** [reply]
Someone ought to initiate a `Ada_202y` (or something..) package tree analogus to the `Ada` package tree.

```Ada
package Std renames Ada; -- Or CLA or mgrojo or something or your pet project
with Std.Text_IO;
```

**Reply by stcarrez** [reply]
> If `Ada.Text_IO` is able to output Latin-1 in a UTF-8 environment, I don't see a reason why `Ada.Directories` is not returning Latin-1 from that UTF-8 environment. OK, there will be a penalty from doing the translation back and forth, when compared to the default mode, which does nothing. But once you've assumed that `Ada.Text_IO` has to convert from Latin-1 due to the environment working in UTF-8, the opposite direction should do the opposite conversion, IMHO.
> 

After looking more closely at [A.16 The Package Directories](http://www.ada-auth.org/standards/12aarm/html/AA-A-16.html), the interpretation of file names and directory names is implementation defined (see 46.a/2).  It is best if the library does not make the UTF-8 to Latin-1 decoding as this could raise some exception with some file names.

Having said that and knowing that Ada.Text_IO is only Latin-1, my previous example is wrong and the following better handles UTF-8 file names (**IF** the file system is using UTF-8).

```Ada
with Ada.Directories;
with Ada.Wide_Wide_Text_IO;
with Ada.Strings.UTF_Encoding.Wide_Wide_Strings;
procedure Readdir is
   use Ada.Wide_Wide_Text_IO;
   use Ada.Strings.UTF_Encoding.Wide_Wide_Strings;
   Search_Filter : constant Ada.Directories.Filter_Type
     := (Ada.Directories.Ordinary_File => True,
         Ada.Directories.Directory     => True,
         Ada.Directories.Special_File  => False);
   Search   : Ada.Directories.Search_Type;
   Ent      : Ada.Directories.Directory_Entry_Type;
begin
   Ada.Directories.Start_Search (Search, Directory => ".",
                                 Pattern => "*", Filter => Search_Filter);
   while Ada.Directories.More_Entries (Search) loop
      Ada.Directories.Get_Next_Entry (Search, Ent);
      Put_Line (Decode (Ada.Directories.Simple_Name (Ent)));
   end loop;
end Readdir;
```

**Reply by mgrojo** [reply]
> After looking more closely at [A.16 The Package Directories](http://www.ada-auth.org/standards/12aarm/html/AA-A-16.html), the interpretation of file names and directory names is implementation defined (see 46.a/2). It is best if the library does not make the UTF-8 to Latin-1 decoding as this could raise some exception with some file names.

I could agree with that, but then why `Ada.Text_IO` is converting Latin-1 to UTF-8 when `-gnatW8` is in place? It should output Latin-1, and the user would have to change the terminal settings to Latin-1 to get clear output. Of course, that is asking too much nowadays, but at least would be consistent with the fact that String and the associated standard packages handle Latin-1. The current behaviour of this switch seems inconsistent because it adds an unexpected (and undocumented, I think) conversion in `Text_IO`.

> 
> Having said that and knowing that Ada.Text_IO is only Latin-1, my previous example is wrong and the following better handles UTF-8 file names (**IF** the file system is using UTF-8).

Would that work on Windows? I guess that, without `Ada.Wide_Wide_Directories`, there is no portable solution. But at least, it's a workaround for UTF-8 systems.

As an example of what I mean. This simple program works mostly right in default mode (except for the length problem in UTF-8 terminal; it works perfect if you configure the terminal in Latin-1, as in the old days). On the other hand, it is miserably broken when using `-gnatW8`. It is not able to give a clear string, no matter if you configure your terminal in UTF-8 or in Latin-1. (Tested with FSF GNAT 12).

```ada
with Ada.Text_IO;

procedure Text_Io_Test is
   Test : constant String := Ada.Text_Io.Get_Line;
begin
   
   Ada.Text_Io.Put_Line (Test);
   Ada.Text_Io.Put_Line ("Length: " & Test'Length'Image);
end Text_Io_Test;
```

`-gnatW8` + UTF-8 terminal:
```
ñéáí@
Ã±Ã©Ã¡Ã­@
Length:  9
```
`-gnatW8` + Latin-1 terminal:
```
$ ./text_io_test
ñéáí@
Ã±Ã©Ã¡Ã­@
Length:  5
```

In default compiler mode + UTF-8 terminal:
```
ñéáí@
ñéáí@
Length:  9
```
In default compiler mode + Latin-1 terminal:
```
ñéáí@
ñéáí@
Length:  5
```

**Reply by mgrojo** [reply]
Discussion in comp.lang.ada related to what I'm saying: https://groups.google.com/g/comp.lang.ada/c/01RzkosroZs

**Reply by stcarrez** [reply]
The issue comes from `Ada.Text_IO.Get_Line` which returns the string as a sequence of bytes (hence UTF-8 if your terminal is UTF-8). If you pass it through `Ada.Strings.UTF_Encoding.Strings.Decode` this will be correct. It looks weird that the Latin-1 <-> UTF-8 encoding/decoding is only made in one direction.

**Reply by mgrojo** [reply]
> It looks weird that the Latin-1 <-> UTF-8 encoding/decoding is only made in one direction.

That's the point. It should be everywhere or nowhere, and it should be independent on how the source code text encoding is selected. That's why I think there should be two switches.

**Reply by mosteo** [reply]
`Get_Line` not returning Latin1 seems like a bug. I'd be curious to see what happens if IO is performed to/from a file instead of the terminal. It's clear that a file generated in some mode should be readable under that mode. And for `-gnatW8` I'd expect to see Latin1 in the Ada side and UTF8 in the file.

The thing is that for files there's also the `Form` parameter which gives even more control. I seem to remember that the same can be achieved with environment variables, which maybe also applies to the standard streams.

And then there's the `-W` switch [for the binder](https://gcc.gnu.org/onlinedocs/gnat_ugn/Switches-for-gnatbind.html) that also alters behavior, but good look with the documentation. Still, I think it only moves trouble around.

**Comment by jquorning** [comment]
I guess program text encoding is best stored as program text meta data in the alire crate.

**Comment by reznikmm** [comment]
Unfortunately this won't work with current gnat compiler, because it reads
all sources in a single  encoding. For instance if you specify -gnatW8 for
your library then compiler will read all dependency as utf8.

**Reply by jquorning** [reply]
чт, 9 бер. 2023, 05:38 користувач Jesper Quorning ***@***.***>
пише:
[…](https://github.com/alire-project/alire/discussions/1334#)

 I guess program encoding is best stored as program text meta data in the
 alire crate.

 —
 Reply to this email directly, view it on GitHub
 <[#1334 (comment)](https://github.com/alire-project/alire/discussions/1334#discussioncomment-5249054)>,
 or unsubscribe
 <https://github.com/notifications/unsubscribe-auth/ABRB775PNX3LBG7PBTCLQ73W3FGFDANCNFSM6AAAAAAVFTYQ4U>
 .
 You are receiving this because you were mentioned.Message ID:
 ***@***.***>
-------------
I guess **M%dollar** solved this by making **UTF-16** the standard..
Most desktop machines are at least **16bit** anyway these days..

**Comment by mosteo** [comment]
Lots of interesting discussion. It reinforces my impressions though that without mandated `-gnatW8` there can't never be peace.

The hard sad truth seems to be that without `-gnatW8`, GNAT is not ARM compliant, and it's working with UTF-8 + `Text_IO` by (bad) chance. Using `-gnatW8` requires programmers to know that `Text_IO` is *not* to be used to output UTF-8.

I don't see how moving forward without `-gnatW8` can be good in the long run. Better that everyone realizes ASAP that `String` is Latin-1 and `Text_IO` should have been named `Latin_1_IO`. 

Of course you can use String to store UTF-8... as long as you deal with the IO in some raw form.

EDIT: and let's not forget that even without `-gnatW8`, a single BOM throws everything into UTF8 mode.

**Comment by jquorning** [comment]
Is there an 'easy way' to cofigure which `standard` to use..?

I read somewhere about a person (from Finland, Hakkinen? perhaps) who(m?) wrote his own standard with - among other things - another definition of `String` for better performance (as I recall it for embedded targets) ..

Update:
He redefined `string` to be `wide_string`, I think it was..
It was - as I recall it - on *comp.lang.ada*.
Perhaps it is @nholsti - http://www.tidorum.fi/en/index.html

**Reply by mgrojo** [reply]
What you say reminds me of Drake, but no mention of embedded and not from Finland, but from Japan. Drake is a different runtime for GNAT, which assumes UTF-8 for String instead of Latin-1: https://github.com/ytomino/drake/wiki/Incompatibility#string-is-utf-8-not-latin-1
